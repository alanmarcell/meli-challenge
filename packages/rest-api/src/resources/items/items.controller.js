import { itemsDto, itemDto } from './items.dto';
import { getItemsService, getItemService, getItemDescriptionService } from './items.datasource';

const errorHandler = (res, error) => res.status(500).json({ message: error.message });

const getItems = async (req, res, _next) => {
  try {
    const response = await getItemsService(req.query.q);
    res.json(itemsDto(response.data));
  } catch (error) {
    errorHandler(res, error);
  }
};

const getItem = async (req, res, _next) => {
  try {
    const { id } = req.params;
    const itemResponse = await getItemService(id);
    const itemDescriptionResponse = await getItemDescriptionService(id);
    res.json(itemDto(itemResponse.data, itemDescriptionResponse.data));
  } catch (error) {
    errorHandler(res, error);
  }
};

export { getItems, getItem };
