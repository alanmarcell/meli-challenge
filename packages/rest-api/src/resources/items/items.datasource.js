import axios from 'axios';

const instance = axios.create({
  baseURL: 'https://api.mercadolibre.com',
});

const getItemsService = (query) => instance.get(`/sites/MLA/search?q=${query}`);
const getItemService = (id) => instance.get(`/items/${id}`);
const getItemDescriptionService = (id) => instance.get(`/items/${id}/description`);

export {
  instance,
  getItemsService,
  getItemService,
  getItemDescriptionService,
};
