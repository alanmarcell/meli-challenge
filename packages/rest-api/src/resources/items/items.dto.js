import * as R from 'ramda';

const getCategories = (itemsRaw) => R.map((value) => value.name,
  R.prop('values',
    R.find((availableFilter) => availableFilter.id === 'category',
      R.prop('available_filters', itemsRaw))));

const itemsDto = (itemsRaw) => {
  const categories = getCategories(itemsRaw);

  const items = R.map((result) => {
    const amount = R.split('.', result.price.toString());

    return {
      id: result.id,
      title: result.title,
      price: {
        currency: result.currency_id,
        amount: Number(amount[0]),
        decimals: amount[1] ? Number(amount[1]) : 0,
      },
      picture: result.thumbnail,
      condition: result.condition,
      free_shipping: result.shipping.free_shipping,
    };
  }, R.prop('results', itemsRaw));

  return {
    author: {
      name: 'Alan',
      lastname: 'Marcell',
    },
    categories,
    items,
  };
};

export { itemsDto };
