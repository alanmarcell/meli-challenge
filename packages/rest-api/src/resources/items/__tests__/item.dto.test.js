import { itemDto } from '../items.dto';
import rawItemMock from './item.mock';
import rawItemDescriptionMock from './itemDescription.mock';

describe('Item DTO', () => {
  it('should have correct format', () => {
    expect(itemDto(rawItemMock, rawItemDescriptionMock)).toMatchSnapshot();
  });
});
