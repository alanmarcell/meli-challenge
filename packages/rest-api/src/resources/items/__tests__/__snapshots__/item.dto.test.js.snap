// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item DTO should have correct format 1`] = `
Object {
  "author": Object {
    "lastname": "Marcell",
    "name": "Alan",
  },
  "item": Object {
    "condition": "new",
    "description": "AMSTERDAM IMPORTADOR DIRECTO

MERCADO LIDER PLATINUM
Contamos con el 100% de calificaciones positivas y mas de 8 años de experiencia en ventas online! Tu compra es garantizada! 

PRODUCTO: CONSOLA DE VIDEO JUEGO FAMILYALIEN 1

CARACTERISTICAS: 

-FamilyGameAlien Pro 1 + 1juego de 8 bits +2 Joys+pistola
-Excelente consola family game con forma de playstation
-Transformador a 220 v
- 2 joysticks
-Juegos de regalo (incorporado)
-Pistola (no es compatible con lcd)
-Cables de conexión (audio y video)

Realiza tu compra de la formas mas segura, mediante MERCADOPAGO. Consulta promociones disponibles. 

FORMAS DE ENVIO:
- MercadoEnvios. mediante Correo Argentino. Podes calcular el costo de envío debajo del precio de la publicación. 
- Micro a terminal: A todo el país con pago en destino.
- Micro con entrega a domicilio (solo CABA)
- Retiro personalmente: Con DNI en Mar del Plata

¿QUIÉNES SOMOS?
AMSTERDAM IMPORTADOR DIRECTO es una empresa argentina con más de 8 años de trayectoria y experiencia en el mercado, brindando asesoramiento personalizado al momento de la compra. Trabajamos todos los días para lograr que tengas la mejor experiencia de compra. Siempre recordando lo esencial, el conocimiento de lo que vendemos, y el respeto, honrando la confianza que depositasen nosotros con cada compra.

¿PORQUÉ COMPRAR CON NOSOTROS?
Siempre buscamos el modo de estar cerca de nuestros clientes.Nuestra atención es personalizada, con vendedores especializados que te asesoraran en cada producto. Contamos con local, por lo que además de adquirir un producto nuevo, sellado y 100% original, podés optar por retirarlo personalmente.",
    "free_shipping": false,
    "id": "MLA653546463",
    "picture": "http://mla-s2-p.mlstatic.com/717025-MLA25351455547_022017-I.jpg",
    "price": Object {
      "amount": 1995,
      "currency": "ARS",
      "decimals": 85,
    },
    "sold_quantity": 3,
    "title": "Consola De Video Juego Family Alien 1",
  },
}
`;
